processor/src/chiplets/hasher/mod.rs

```diff
@@ -139,74 +222,127 @@ impl Hasher {
+
+        let lookups = self.get_last_lookups(init_lookup_idx);
+        (addr, root, lookups)
     }
 
-    /// Performs Merkle root update computation and records its execution trace.
+    /// Performs Merkle root update computation and records its execution trace, as well as the
+    /// lookups required for verifying its correctness so that they can be provided to the Chiplets
+    /// Bus when the trace is finalized.
     ///
     /// The computation consists of two Merkle path verification procedures for a node at the
     /// specified index. The procedures compute Merkle roots for the specified path for the old
     /// value of the node (value before the update), and the new value of the node (value after
     /// the update).
     ///
     /// The returned tuple contains these roots, as well as the row address of the execution trace
-    /// at which the computation started.
+    /// at which the computation started and the lookups required to verify the computation so that
+    /// the correct requests can be sent by the caller to the Chiplets Bus.
     ///
     /// # Panics
     /// Panics if:
     /// - The provided path does not contain any nodes.
     /// - The provided index is out of range for the specified path.
-    pub fn update_merkle_root(
+    pub(super) fn update_merkle_root(
         &mut self,
         old_value: Word,
         new_value: Word,
         path: &[Word],
         index: Felt,
-    ) -> (Felt, Word, Word) {
+    ) -> (Felt, Word, Word, &[HasherLookup]) {
         let addr = self.trace.next_row_addr();
+        let init_lookup_idx = self.next_lookup_idx();
         let index = index.as_int();
+
         let old_root =
             self.verify_merkle_path(old_value, path, index, MerklePathContext::MrUpdateOld);
         let new_root =
             self.verify_merkle_path(new_value, path, index, MerklePathContext::MrUpdateNew);
-        (addr, old_root, new_root)
+
+        let lookups = self.get_last_lookups(init_lookup_idx);
+        (addr, old_root, new_root, lookups)
     }
 
     // TRACE GENERATION
     // --------------------------------------------------------------------------------------------
 
-    /// Fills the provided trace fragment with trace data from this hasher trace instance. This
-    /// also returns the trace builder for hasher-related auxiliary trace columns.
-    pub fn fill_trace(self, trace: &mut TraceFragment) -> AuxTraceBuilder {
+    /// Fills the provided trace fragment with trace data from this hasher trace instance and sends
+    /// all hasher lookups to the ChipletsBus. This also returns the trace builder for
+    /// hasher-related auxiliary trace columns.
+    pub(super) fn fill_trace(
+        self,
+        trace: &mut TraceFragment,
+        chiplets_bus: &mut ChipletsBus,
+    ) -> AuxTraceBuilder {
+        // provide all lookups to the ChipletsBus.
+        for lookup in self.lookups {
+            chiplets_bus.provide_hasher_lookup(lookup, lookup.cycle());
+        }
```

Comment:
We could do this more efficiently than looping through, but it would require more changes to the chiplets bus and it would make this approach inconsistent with the other chiplets, so I wanted to start with something that was simple, clean, and easy to reason about (at least until all tests were passing).

I could change this now or add a TODO and address it in a future PR. There are two approaches that come to mind which we could take:
1. track all the lookups (like now) and then add them in bulk at this point.
2. provide lookups to the bus when they're first created instead of recording them here.

Reply:
note: along with other performance optimizations, we'll leave this for the future